python/tank/bootstrap/threading.py

```diff
@@ -0,0 +1,269 @@
+        # where the main event loop can handle its deletion.
+        self._worker.moveToThread(self.thread())
+
+        if self._failed_callback:
+            # Handle cleanup after failed completion of the toolkit bootstrap.
+            self._failed_callback("sgtk", exception)
+
+
+class BootstrapWorker(QtCore.QObject):
+    """
+    Bootstrap worker that can bootstrap an :class:`sgtk.Sgtk` instance asynchronously in a background thread.
+    """
+
+    # Qt signal emitted while the bootstrap worker is progressing in its work in the background.
+    progressing = QtCore.Signal(str, int, int)
+
+    # Qt signal emitted when the bootstrap worker successfully completes its work in the background.
+    completed = QtCore.Signal(Sgtk)
+
+    # Qt signal emitted when the bootstrap worker fails to complete its work in the background.
+    failed = QtCore.Signal(Exception)
+
+    def __init__(self, toolkit_manager, engine_name, entity):
+        """
+        Initializes an instance of the bootstrap worker.
+
+        :param toolkit_manager: :class:`sgtk.bootstrap.ToolkitManager` instance bootstrapping the engine.
+        :param engine_name: Name of the engine used to resolve a configuration.
+        :param entity: Shotgun entity used to resolve a project context.
+        """
+
+        super(BootstrapWorker, self).__init__()
+
+        self._toolkit_manager = toolkit_manager
+        self._engine_name = engine_name
+        self._entity = entity
+
+    @QtCore.Slot()
+    def work(self):
+        """
+        Bootstraps a toolkit instance for the given engine and entity and
+        signal the progress, and the completion or the failure of this work.
+        """
+
+        try:
+
+            # Install the bootstrap progress reporting callback.
+            self._toolkit_manager.set_progress_callback(self._report_progress)
+
+            # Bootstrap a toolkit instance for the given engine and entity.
+            toolkit = self._toolkit_manager.bootstrap_sgtk(self._engine_name, self._entity)
+
+            # Signal completion of the toolkit bootstrap.
+            self.completed.emit(toolkit)
+
+        except Exception as exception:
+
+            # Signal failure of the toolkit bootstrap.
+            self.failed.emit(exception)
+
+        finally:
+
+            # Remove the bootstrap progress reporting callback.
+            self._toolkit_manager.set_progress_callback(None)
+
+    def _report_progress(self, message, current_index, maximum_index):
+        """
+        Callback function that reports back on the toolkit bootstrap progress.
+
+        :param message: Progress message to report.
+        :param current_index: Optional current item number being looped over.
+        :param maximum_index: Optional maximum item number being looped over.
+        """
+
+        # Signal the toolkit bootstrap progress.
+        self.progressing.emit(message, current_index, maximum_index)
+
+
+def get_thread_info_msg(caller):
```

Comment:
Is this more of a debugging method? Maybe we can make it protected and mark it as such?